File: python/1306.jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        q = deque()
        q.append(start)
        while len(q):
            current = q.popleft()
            if arr[current] == 0:
                return True
            if arr[current] < 0:
                continue
            arr[current] = -arr[current]
            for _next in [current+arr[current], current-arr[current]]:
                if _next < 0 or _next >= len(arr) or arr[_next] < 0:
                    continue
                q.append(_next)
        return False

File: cpp/1306.jump-game-iii.cpp
class Solution {
public:
    bool solve1(vector<int>& arr, int start) {
	    int n = arr.size();
	    queue<int> q;
	    q.push(start);
	    arr[start]=-arr[start];
	    while(!q.empty()){
		    auto current = q.front(); q.pop();
		    if(arr[current] == 0)return true;
	            vector<int> next_all = {current+arr[current], current-arr[current]};
		    for(auto next:next_all){
			    if(next<0 || next>=n)continue;
			    if(arr[next]<0)continue;
			    arr[next] = -arr[next];
			    q.push(next);
		    }
	    }
	    return false;
    }
    bool canReach(vector<int>& arr, int start) {
	    return solve1(arr, start);
    }
};
